### ISSUE TITLE
Still MIDI problems with notes off (Orion / FLStudio)

### ISSUE THREAD

**Issue by p0nce** [AuburnSounds/Dplug#575]
Follow-up to badly fixed #539

Customer says:
> My Graillon 2 keeps crashing on my fl studio
by crashing i mean the autotuned vocals became wobbly and inaccurate
it happened a lot of times and it's really frustrating please fix this bug

For now the bug is only reproduced in a realistic session with many instances.
In isolation with only a MIDI track and audio track, no such bug.

Likely explanations: we don't handle all types of "All Notes Off", there are 5 CC to do that. (EDIT: NO)
```
c = 123, v = 0: All Notes Off (See text for description of actual mode commands.)
c = 124, v = 0: Omni Mode Off
c = 125, v = 0: Omni Mode On
c = 126, v = M: Mono Mode On (Poly Off) where M is the number of channels (Omni Off) or 0 (Omni On)
c = 127, v = 0: Poly Mode On
```

Also possibly related to "Mono Mode", where Notes Off are implicit? (EDIT: NO)

**Comment by p0nce** [comment]
Annoying to reproduce, doesn't happen anymore if we simplify the session.
Maybe related to our VST2 MIDI queue?

**Comment by p0nce** [comment]
Output of Note ON/OFF sequence:
```
NOTE ON = 57
  _currentNumberOfNotePlayed = 1
NOTE OFF = 57
  _currentNumberOfNotePlayed = 0
NOTE ON = 63
  _currentNumberOfNotePlayed = 1
NOTE OFF = 63
  _currentNumberOfNotePlayed = 0
NOTE ON = 65
  _currentNumberOfNotePlayed = 1
NOTE ON = 67
  _currentNumberOfNotePlayed = 2
NOTE ON = 68
  _currentNumberOfNotePlayed = 3
NOTE OFF = 65
  _currentNumberOfNotePlayed = 2
NOTE ON = 69
  _currentNumberOfNotePlayed = 3
NOTE OFF = 67
  _currentNumberOfNotePlayed = 2
NOTE ON = 71
  _currentNumberOfNotePlayed = 3
NOTE ON = 70
  _currentNumberOfNotePlayed = 4
NOTE OFF = 68
  _currentNumberOfNotePlayed = 3
NOTE OFF = 69
  _currentNumberOfNotePlayed = 2
NOTE OFF = 70
  _currentNumberOfNotePlayed = 1
NOTE OFF = 71
  _currentNumberOfNotePlayed = 0
NOTE ON = 105
  _currentNumberOfNotePlayed = 1
NOTE ON = 105
  _currentNumberOfNotePlayed = 2
NOTE OFF = 105
  _currentNumberOfNotePlayed = 1
NOTE OFF = 105
  _currentNumberOfNotePlayed = 0
NOTE ON = 105
  _currentNumberOfNotePlayed = 1
NOTE ON = 105
  _currentNumberOfNotePlayed = 2
NOTE OFF = 105
  _currentNumberOfNotePlayed = 1
NOTE OFF = 105
  _currentNumberOfNotePlayed = 0
NOTE ON = 103
  _currentNumberOfNotePlayed = 1
NOTE ON = 103
  _currentNumberOfNotePlayed = 2
NOTE OFF = 103
  _currentNumberOfNotePlayed = 1
NOTE OFF = 103
  _currentNumberOfNotePlayed = 0
NOTE ON = 103
  _currentNumberOfNotePlayed = 1
NOTE ON = 103
  _currentNumberOfNotePlayed = 2
NOTE OFF = 103
  _currentNumberOfNotePlayed = 1
NOTE OFF = 103
  _currentNumberOfNotePlayed = 0
NOTE ON = 101
  _currentNumberOfNotePlayed = 1
NOTE ON = 101
  _currentNumberOfNotePlayed = 2
NOTE OFF = 101
  _currentNumberOfNotePlayed = 1
NOTE OFF = 101
  _currentNumberOfNotePlayed = 0
NOTE ON = 99
  _currentNumberOfNotePlayed = 1
NOTE ON = 99
  _currentNumberOfNotePlayed = 2
NOTE OFF = 99
  _currentNumberOfNotePlayed = 1
NOTE ON = 98
  _currentNumberOfNotePlayed = 2
NOTE ON = 98
  _currentNumberOfNotePlayed = 3
NOTE OFF = 99
  _currentNumberOfNotePlayed = 2
NOTE OFF = 98
  _currentNumberOfNotePlayed = 1
NOTE OFF = 98
  _currentNumberOfNotePlayed = 0
NOTE ON = 96
  _currentNumberOfNotePlayed = 1
NOTE ON = 96
  _currentNumberOfNotePlayed = 2
NOTE OFF = 96
  _currentNumberOfNotePlayed = 1
NOTE OFF = 96
  _currentNumberOfNotePlayed = 0
NOTE ON = 95
  _currentNumberOfNotePlayed = 1
NOTE ON = 95
  _currentNumberOfNotePlayed = 2
NOTE OFF = 95
  _currentNumberOfNotePlayed = 1
NOTE OFF = 95
  _currentNumberOfNotePlayed = 0
NOTE ON = 92
  _currentNumberOfNotePlayed = 1
NOTE ON = 92
  _currentNumberOfNotePlayed = 2
NOTE OFF = 92
  _currentNumberOfNotePlayed = 1
NOTE OFF = 92
  _currentNumberOfNotePlayed = 0
NOTE ON = 91
  _currentNumberOfNotePlayed = 1
NOTE OFF = 91
  _currentNumberOfNotePlayed = 0
NOTE OFF = 91
  _currentNumberOfNotePlayed = 0
NOTE ON = 91
  _currentNumberOfNotePlayed = 1
```

**Comment by p0nce** [comment]
The host sends 
```
Note On 102
Note On 102 
Note Off 102
Note Off 102
```
all with the same timestamps.
In order to prevent timestamp inversion (can't trust the host!), the messages are diligently put into a priority-queue (min heap) that _isn't stable_.
The messages get inverted as follow:
```
Note On 102
Note Off 102
Note Off 102 // this one is ignored since no note is currently playing!
Note On 102
```
That ends up creating the problem I was supposed to avoid.

**Comment by p0nce** [comment]
Fixed in v12.0.10
